velox/exec/TopNRowNumber.h

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/exec/HashTable.h"
+#include "velox/exec/Operator.h"
+
+namespace facebook::velox::exec {
+
+/// Partitions the input using specified partitioning keys, sorts rows within
+/// partitions using specified sorting keys, assigns row numbers and returns up
+/// to specified number of rows per partition.
+///
+/// It is allowed to not specify partitioning keys. In this case the whole input
+/// is treated as a single partition.
+///
+/// At least one sorting key must be specified.
+///
+/// The limit (maximum number of rows to return per partition) must be greater
+/// than zero.
+///
+/// This is an optimized version of a Window operator with a single row_number
+/// window function followed by a row_number <= N filter.
+class TopNRowNumber : public Operator {
+ public:
+  TopNRowNumber(
+      int32_t operatorId,
+      DriverCtx* driverCtx,
+      const std::shared_ptr<const core::TopNRowNumberNode>& node);
+
+  bool needsInput() const override {
+    return true;
+  }
+
+  void addInput(RowVectorPtr input) override;
+
+  RowVectorPtr getOutput() override;
+
+  BlockingReason isBlocked(ContinueFuture* /*future*/) override {
+    return BlockingReason::kNotBlocked;
+  }
+
+  void noMoreInput() override;
+
+  bool isFinished() override;
+
+ private:
+  /// A priority queue to keep track of top 'limit' rows for a given partition.
+  struct TopRows {
+    struct Compare {
+      RowComparator& comparator;
+
+      bool operator()(const char* lhs, const char* rhs) {
+        return comparator(lhs, rhs);
+      }
+    };
+
+    std::priority_queue<char*, std::vector<char*, StlAllocator<char*>>, Compare>
+        rows;
+
+    TopRows(HashStringAllocator* allocator, RowComparator& comparator)
+        : rows{{comparator}, StlAllocator<char*>(allocator)} {};
+  };
+
+  /// A synthetic Aggregate passed to HashTable and its RowContainer to
+  /// associate a 'TopRows' struct with each partition.
+  ///
+  /// TODO Refactor RowContainer and HashTable to not require an 'Aggregate'
```

Comment:
This is very nice. I had similar ideas for Window with HashTable. However, we need to spill to disk there. The spilling part is simple for such a structure. But on the read side, the merging needs to merge across these containers also.